src/main/java/org/sasanlabs/service/vulnerability/ssrf/SSRFVulnerability.java

```diff
@@ -1,52 +1,221 @@
+        } else {
+            return getGenericVulnerabilityResponseWhenFileName(url);
+        }
+    }
+
+    @AttackVector(
+            vulnerabilityExposed = VulnerabilityType.SSRF_WITH_INPUT_FILTERS,
+            description = "FILE_URL_IF_NOT_FILE_PROTOCOL_AND_LOCALHOST_AND_VARIANTS",
+            payload = "SSRF_PAYLOAD_LEVEL_4")
+    @VulnerableAppRequestMapping(value = LevelConstants.LEVEL_4, htmlTemplate = "LEVEL_1/SSRF")
+    public ResponseEntity<GenericVulnerabilityResponseBean<StringBuilder>>
+            getVulnerablePayloadLevel4(@RequestParam(FILE_URL) String url) throws IOException {
+        if (isUrlValid(url)) {
+            for (String variant : LOCALHOST_VARIANTS_PATTERN) {
+                if (Pattern.matches(variant, url)) {
+                    return new ResponseEntity<>(
+                            new GenericVulnerabilityResponseBean<>(
+                                    new StringBuilder("Variants of localhost are not allowed"),
+                                    false),
+                            HttpStatus.OK);
+                }
+            }
+            return getGenericVulnerabilityResponseWhenFilters(url);
+        } else {
+            return getGenericVulnerabilityResponseWhenFileName(url);
+        }
+    }
+
+    @AttackVector(
+            vulnerabilityExposed = VulnerabilityType.SSRF_WHITE_LIST,
+            description = "FILE_URL_IF_IN_THE_WHITELIST")
+    @VulnerableAppRequestMapping(
+            value = LevelConstants.LEVEL_5,
+            htmlTemplate = "LEVEL_1/SSRF",
+            variant = Variant.SECURE)
+    public ResponseEntity<GenericVulnerabilityResponseBean<StringBuilder>>
+            getVulnerablePayloadLevel5(@RequestParam(FILE_URL) String url) throws IOException {
+        if (isUrlValid(url)) {
+            if (!Arrays.asList(URL_WHITE_LIST).contains(url)) {
+                return new ResponseEntity<>(
+                        new GenericVulnerabilityResponseBean<>(
+                                new StringBuilder("URL not part of the White List"), false),
+                        HttpStatus.OK);
+            }
+            return getGenericVulnerabilityResponseWhenURL(url);
+        } else {
             return new ResponseEntity<>(
-                    new GenericVulnerabilityResponseBean<>(bytes, true), HttpStatus.OK);
-        } catch (Exception e) {
-            LOGGER.error(
-                    "Following exception occurred while opening the connection to {}", urlImage, e);
-        }
-        return new ResponseEntity<>(
-                new GenericVulnerabilityResponseBean<>(
-                        ("Failed to fetch image from URL " + urlImage)
-                                .getBytes(StandardCharset.UTF_8),
-                        false),
-                HttpStatus.BAD_REQUEST);
+                    new GenericVulnerabilityResponseBean<>(
+                            new StringBuilder("URL not valid"), false),
+                    HttpStatus.OK);
```

Comment:
No need for checking if url is valid or not, just compare with whitelist and if it doesn't match return not a valid url.